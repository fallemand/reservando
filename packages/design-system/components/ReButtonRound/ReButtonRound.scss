@import "../../styles/variables.scss";
@import "../ReButton/buttons.scss";

.re-button-round {
  font-weight: $font-weight-bold;
  cursor: pointer;
  font-family: inherit;
  padding: 1em;
  text-decoration: none;
  border-radius: 50%;

  &--small {
    font-size: 20px;
    padding: $bdu;
  }

  &--medium {
    font-size: 24px;
    padding: $bdu * 1.5;
  }

  &--large {
    font-size: 24px;
    padding: $bdu * 2;
  }

  &--primary {
    @include cta-generator($primary-500, $primary-600, $primary-700, $black, $white, $grey-300);
  }

  &--secondary {
    @include cta-generator($white, $grey-200, $grey-300, $black, $black, $grey-300, $grey-500);
  }
}
