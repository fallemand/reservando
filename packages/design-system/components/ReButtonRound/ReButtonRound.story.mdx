import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import { ReButtonRound } from "@reservando/design-system";

<Meta
  title="Components/ButtonRound"
  component={ReButtonRound}
  argTypes={{
    size: { control: { type: "select", options: ["small", "medium", "large"] } },
    modifier: { control: { type: "select", options: ["cta-red", "cta-uva", "outline-uva"] } },
    type: { type: "select", options: ["button", "submit"] },
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { ReButtonRound },
  template: '<ReButtonRound :size="size" :modifier="modifier">{{label}}</ReButtonRound>',
});

# ReButtonRound

<Canvas>
  <Story name="basic">
    {{
      components: { ReButtonRound },
      template: `
        <div>
          <ReButtonRound icon="arrow-right" />
        </div>
      `,
    }}
  </Story>
</Canvas>

<ArgsTable of={ReButtonRound} />

## How to use

```html
<template>
  <ReButtonRound @click="clickHandler" icon="arrow-right" />
</template>

<script lang="ts">
  import { ReButtonRound } from "@reservando/design-system";
</script>
```

## Using modifiers

<Canvas>
  <Story name="modifiers">
    {{
      components: { ReButtonRound },
      data() {
        return {
          checked: true,
        };
      },
      template: `
      <div>
        <ReButtonRound modifier="primary" class="mr-2" icon="arrow-right" />
        <ReButtonRound modifier="secondary" class="mr-2" icon="arrow-right" />
        <p class="hint mt-4">Disabled</p>
        <ReButtonRound modifier="primary" disabled="true" class="mr-2" icon="arrow-right" />
        <ReButtonRound modifier="secondary" disabled="true" class="mr-2" icon="arrow-right" />
      </div>
    `,
    }}
  </Story>
</Canvas>

## Using sizes

<Canvas>
  <Story name="sizes">
    {{
      components: { ReButtonRound },
      data() {
        return {
          checked: true,
        };
      },
      template: `
      <div>
        <ReButtonRound size="small" class="mr-2" icon="arrow-right" />
        <ReButtonRound size="medium" class="mr-2" icon="arrow-right" />
        <ReButtonRound size="large" class="mr-2" icon="arrow-right" />
      </div>
    `,
    }}
  </Story>
</Canvas>
