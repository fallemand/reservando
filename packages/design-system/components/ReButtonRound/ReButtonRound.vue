<template>
  <button
    :class="[
      're-button-round',
      `re-button-round--${size}`,
      `re-button-round--${modifier}`,
    ]"
    type="button"
  >
    <ReIcon :name="icon" />
  </button>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";
import ReIcon from "../ReIcon/ReIcon.vue";

export enum ReButtonRoundSize {
  small = "small",
  medium = "medium",
  large = "large",
}
export enum ReButtonRoundModifier {
  primary = "primary",
  secondary = "secondary",
}

const ReButtonRound = defineComponent({
  components: { ReIcon },
  name: "ReButtonRound",
  props: {
    size: {
      default: "medium",
      type: String as PropType<ReButtonRoundSize>,
      validator: (value: string): boolean =>
        Object.values(ReButtonRoundSize).includes(value as ReButtonRoundSize),
    },
    modifier: {
      default: "primary",
      type: String as PropType<ReButtonRoundModifier>,
      validator: (value: string): boolean =>
        Object.values(ReButtonRoundModifier).includes(value as ReButtonRoundModifier),
    },
    icon: {
      required: true,
      type: String,
    },
  },
});

export default ReButtonRound;
</script>

<style lang="scss" src="./ReButtonRound.scss" />
