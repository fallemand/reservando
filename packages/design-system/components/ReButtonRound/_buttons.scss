// -------------------------------------------------------------------------------
//  Buttons
//
// üíÅ‚Äç use variables rather than hard-coded pixel/REM values
// -------------------------------------------------------------------------------

$btn-border-width: 1px;
$btn-border-style: solid;

/**
* Cta styles
*/
@mixin cta-generator($color, $hover, $press, $focus, $font, $disabled, $border: transparent) {
  color: $font;
  background: $color;
  @if $border {
    border-color: $border;
  } @else {
    border-color: $color;
  }
  border-style: $btn-border-style;
  border-radius: $bdu;
  border-width: $btn-border-width;
  transition: background-color 0.2s ease-out;
  outline: 0;

  &:hover {
    background: $hover;
    @if $border {
      border-color: $border;
    } @else {
      border-color: $hover;
    }
  }

  &:active {
    background: $press;
    @if $border {
      border-color: $border;
    } @else {
      border-color: press;
    }
  }

  &:disabled {
    color: $grey-600;
    background: $disabled;
    border-color: $grey-300;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    color: $font;
    background: $color;
    border-color: $focus;
  }
}

/**
* Link styles
*/
@mixin link-generator($color, $hover, $press, $focus) {
  text-decoration: none;
  background: transparent;
  border: 0;
  color: $color;
  transition: color 0.2s ease-out;
  padding: 0.25em 0.5em;
  outline: 0;

  &:hover {
    color: $hover;
  }

  &:active {
    color: $press;
  }

  &:disabled {
    color: $grey-600;
    background: transparent;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    color: $focus;
  }
}
