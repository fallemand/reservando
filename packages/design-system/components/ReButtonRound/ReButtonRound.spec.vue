<script lang="ts">
import { shallowMount } from "@vue/test-utils";
import ReButtonRound, { ReButtonRoundSize } from "./ReButtonRound.vue";

describe("ReButtonRound component", () => {
  const props = {
    icon: "__ICON__",
    size: ReButtonRoundSize.large,
  };
  it("should render with default props", () => {
    const wrapper = shallowMount(ReButtonRound, {
      props,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render disabled button", () => {
    const wrapper = shallowMount(ReButtonRound, {
      props,
      attrs: {
        disabled: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should bind all listeners to the button", () => {
    const listeners = {
      onClick: jest.fn(),
      onMouseleave: jest.fn(),
      onMouseover: jest.fn(),
    };
    const wrapper = shallowMount(ReButtonRound, {
      props,
      attrs: listeners,
    });
    const input = wrapper.find("button");
    input.trigger("click");
    input.trigger("mouseleave");
    input.trigger("mouseover");
    expect(listeners.onClick).toHaveBeenCalled();
    expect(listeners.onMouseleave).toHaveBeenCalled();
    expect(listeners.onMouseover).toHaveBeenCalled();
  });
});
</script>
