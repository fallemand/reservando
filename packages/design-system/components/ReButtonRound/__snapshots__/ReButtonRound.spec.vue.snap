// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReButtonRound component should render disabled button 1`] = `
<button
  class="re-button-round re-button-round--large re-button-round--primary"
  disabled=""
  type="button"
>
  <re-icon-stub
    name="__ICON__"
  />
</button>
`;

exports[`ReButtonRound component should render with default props 1`] = `
<button
  class="re-button-round re-button-round--large re-button-round--primary"
  type="button"
>
  <re-icon-stub
    name="__ICON__"
  />
</button>
`;
